### PR TITLE
Improve support for Apollo Federation 2.0

### LINKED ISSUE #1827: Apollo Federation 2
Since Apollo Federation 2 is now released and introduced a few new directives and some optional fields, is there any current efforts in the project to support those?

What is new in Federation 2: https://www.apollographql.com/docs/federation/federation-2/new-in-federation-2

Compatibility Chart: https://github.com/apollographql/apollo-federation-subgraph-compatibility

### PR DESCRIPTION
**Pull request by patrick91**
## Description

While looking at https://github.com/strawberry-graphql/strawberry/issues/2044 I found some issues with our support for apollo federation, this PR addresses that, it also improves Mypy and Pyright support

- [x] Split up tests into multiple files
- [x] Add a link directive
- [x] Add federation directives support for input and interface
- [x] Automatically add the link directive on the schema based on usage
- [x] Decide if we should keep printing the directives (we do that for v1)
- [x] Add opt-in into v2
- [x] check if we have all the directives and that they are defined correctly
- [x] docs
- [x] federation support for scalars, enums, enum values and unions

Closes #1827

## Types of Changes

- [ ] Core
- [x] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Documentation

### CODE REVIEW

**Comment by botberry** [comment]
# Pre-release
:wave:

Releasing commit [911f520da827e43954d57e9b3bfad8ca605d5f5b] to PyPi as pre-release! :package:

**Comment by botberry** [comment]
Thanks for adding the `RELEASE.md` file!



Here's a preview of the changelog:

---

This release adds full support for Apollo Federation 2.0. To opt-in you need to
pass `enable_federation_2=True` to `strawberry.federation.Schema`, like in the
following example:

```python
@strawberry.federation.type(keys=["id"])
class User:
    id: strawberry.ID

@strawberry.type
class Query:
    user: User

schema = strawberry.federation.Schema(query=Query, enable_federation_2=True)
```

This release also improves type checker support for the federation.


---
Here's the preview release card for twitter:




Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to @patrick91 for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

File: docs/guides/federation.md [hunk 1]
```diff
@@ -138,50 +181,87 @@ the requested `id` and a fixed number of reviews.
...
+In two terminal windows, run the following commands:
+
+```bash
+cd books
+strawberry server --port 3500 app
+```
+
+```bash
+cd reviews
+strawberry server --port 300 app
```
Comment (atatsu) [inline docs/guides/federation.md:206]: I was reading through the docs preview (awesome work by the way with this PR -- and strawberry in general) and noticed this typo: `300` => `3000`.

**Pull request merged**